### PR TITLE
Optional page

### PR DESCRIPTION
**Pull request by Sakshi-gupta-GL**
optional page and customer changes

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -23,10 +23,15 @@ const App = () => {
       limitCount: "",
       expiration: "",
     },
+    optional: {
```
Comment (pinak-globallogic) [inline src/App.jsx:26]: Optional comment and amount of keys are two different things. Take properties outside of optional object.

**Review by pinak-globallogic — CHANGES_REQUESTED** [review]

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 2]
```diff
@@ -1,12 +1,60 @@
...
 
+const customerNameList = [
+  { name: "Datacard LLC" },
+  { name: "Entrust Datacard Inc." },
+  { name: "Datacard Germany GmbH" },
+];
+
+const filter = createFilterOptions();
+
+var newCustomerAddedMsg;
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/CustomerDetails.jsx:18]: Remove

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 3]
```diff
@@ -1,12 +1,60 @@
+import React from "react";
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/CustomerDetails.jsx:1]: Separate groups for react, material ui and internal components. Put line breaks between each group

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 4]
```diff
@@ -1,12 +1,60 @@
...
 import Grid from "@mui/material/Grid";
 import Typography from "@mui/material/Typography";
 import TextField from "@mui/material/TextField";
+import Autocomplete, { createFilterOptions } from "@mui/material/Autocomplete";
 import { useContext } from "react";
 import { CustomCard, CustomCardContent } from "./GenerateLicense";
 import { AppContext } from "App";
 
+const customerNameList = [
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/CustomerDetails.jsx:10]: Add TODO comment that hard coded data should be removed once APIs are available

File: src/Pages/License/Generate/GenerateLicense.jsx [hunk 5]
```diff
@@ -4,11 +4,14 @@ import { styled } from "@mui/styles";
 import CardContent from "@mui/material/CardContent";
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/GenerateLicense.jsx:4]: Separate groups for react, material ui and internal components. Put line breaks between each group

File: src/Pages/License/Generate/GenerateLicense.jsx [hunk 6]
```diff
@@ -4,11 +4,14 @@ import { styled } from "@mui/styles";
...
+import { useState, useContext } from "react";
+import { useNavigate } from "react-router-dom";
+import { AppContext } from "App";
 import CustomerDetails from "./CustomerDetails";
 import ProductDetails from "./ProductDetails";
 import FeatureDetails from "./FeatureDetails";
 import LimitationDetails from "./LimitationDetails";
+import OptionalDetails from "./OptionalDetails";
 
 export const CustomCard = styled(Card)(() => ({
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/GenerateLicense.jsx:16]: If these are used by some other component also, move them to a utility file. This does not look related to Generate License

File: src/Pages/License/Generate/GenerateLicense.jsx [hunk 7]
```diff
@@ -32,30 +35,57 @@ const renderSwitch = (count) => {
...
-  
+  const navigate = useNavigate();
+  const { state } = useContext(AppContext);
+  const btnDisabled = state.error;
+
   return (
     <>
       <Grid item container direction="column" {...props}>
         {detailsPage}
         <Stack direction="row" justifyContent="space-between">
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/GenerateLicense.jsx:61]: Use row-reverse

File: src/Pages/License/Generate/GenerateLicense.jsx [hunk 8]
```diff
@@ -32,30 +35,57 @@ const renderSwitch = (count) => {
...
-          {count > 0 ? (
-            <Button variant="text" onClick={() => setCount(count - 1)}>
+          {count > 0 && count < 5 ? (
+            <Button variant="text" disabled={btnDisabled} onClick={() => {setCount(count - 1)}}>
               Go Back
             </Button>
           ) : (
             <div style={{ display: "block" }} />
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/GenerateLicense.jsx:67]: No need for this in case of row-reverse

File: src/Pages/License/Generate/GenerateLicense.jsx [hunk 9]
```diff
@@ -32,30 +35,57 @@ const renderSwitch = (count) => {
...
           )}
-          {count < 3 && (
-            <Button variant="contained" onClick={() => setCount(count + 1)}>
+          {(count <= 3 && (
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/GenerateLicense.jsx:69]: Add only 1 button for count <=3, 4 and 5. Handle the text and onClick based on count

File: src/Pages/License/Generate/OptionalDetails.jsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
+import React from "react";
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/OptionalDetails.jsx:1]: Separate groups for react, material ui and internal components. Put line breaks between each group

File: src/Pages/License/Generate/OptionalDetails.jsx [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+import Grid from "@mui/material/Grid";
+import Typography from "@mui/material/Typography";
+import TextField from "@mui/material/TextField";
+import { useContext } from "react";
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/OptionalDetails.jsx:5]: Merge with above react import

File: src/Pages/License/Generate/OptionalDetails.jsx [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      });
+    }
+  };
+
+  return (
+    <div>
+      <CustomCard>
+        <CustomCardContent>
+          <Grid item mb={1}>
+            <Typography variant="h6">Optional comment</Typography>
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/OptionalDetails.jsx:37]: Add id to each and every UI element

File: src/Pages/License/Generate/OptionalDetails.jsx [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        </CustomCardContent>
+        <CustomCardContent>
+          <Grid item mb={1}>
+            <Typography variant="h6">Amount of keys</Typography>
+          </Grid>
+          <Grid item mb={2}>
+            <Typography variant="caption">
+              In some instances you might need to generate multiple keys with
+              identical settings. <br />
+              Hence, you can specify amount of key generated. The default value
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/OptionalDetails.jsx:68]: Check the text

File: src/Pages/License/Generate/ProductDetails.jsx [hunk 14]
```diff
@@ -96,6 +96,7 @@ const ProductDetails = () => {
           <Grid item>
             <Autocomplete
               disablePortal
```
Comment (arpitmanglik) [inline src/Pages/License/Generate/ProductDetails.jsx:98]: id for every UI element

File: src/App.jsx (hunk 1 above)
Comment (Sakshi-gupta-GL) [reply src/App.jsx]: done

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 15]
```diff
@@ -1,12 +1,60 @@
+import React from "react";
 import Grid from "@mui/material/Grid";
 import Typography from "@mui/material/Typography";
 import TextField from "@mui/material/TextField";
+import Autocomplete, { createFilterOptions } from "@mui/material/Autocomplete";
 import { useContext } from "react";
```
[the remark this answers is not in the document]
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx]: done

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 16]
```diff
@@ -23,17 +71,57 @@ const CustomerDetails = () => {
...
+                if (option.inputValue) {
+                  return option.inputValue;
+                }
+                // Regular option
+                return option.name;
+              }}
+              sx={{ width: 300 }}
+              renderInput={(params) => (
+                <TextField
+                  {...params}
```
[the remark this answers is not in the document]
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx:100]: id added to all UI elements

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 17]
```diff
@@ -23,17 +71,57 @@ const CustomerDetails = () => {
             </Typography>
```
[the remark this answers is not in the document]
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx:59]: corrected

File: src/Pages/License/Generate/CustomerDetails.jsx [hunk 18]
```diff
@@ -23,17 +71,57 @@ const CustomerDetails = () => {
...
+                  size="small"
+                  required
+                  value={inputValue}
+                />
+              )}
             />
           </Grid>
+          <br />
+          <Grid item mb={2}>
+            {newCustomerAddedMsg}
```
[the remark this answers is not in the document]
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx]: done

File: src/Pages/License/Generate/CustomerDetails.jsx (hunk 2 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx]: done

File: src/Pages/License/Generate/CustomerDetails.jsx (hunk 3 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx]: done for all

File: src/Pages/License/Generate/CustomerDetails.jsx (hunk 4 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/CustomerDetails.jsx:13]: done

File: src/Pages/License/Generate/GenerateLicense.jsx (hunk 5 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/GenerateLicense.jsx]: done

File: src/Pages/License/Generate/GenerateLicense.jsx (hunk 8 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/GenerateLicense.jsx]: removed

File: src/Pages/License/Generate/GenerateLicense.jsx (hunk 9 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/GenerateLicense.jsx]: done

File: src/Pages/License/Generate/OptionalDetails.jsx (hunk 10 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/OptionalDetails.jsx]: done

File: src/Pages/License/Generate/OptionalDetails.jsx (hunk 11 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/OptionalDetails.jsx]: done

File: src/Pages/License/Generate/OptionalDetails.jsx (hunk 13 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/OptionalDetails.jsx]: corrected

File: src/Pages/License/Generate/ProductDetails.jsx [hunk 19]
```diff
@@ -96,6 +96,7 @@ const ProductDetails = () => {
...
+              value={state.product.name +", "+state.product.productItemNo}
```
[the remark this answers is not in the document]
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/ProductDetails.jsx]: done

File: src/Pages/License/Generate/ProductDetails.jsx (hunk 14 above)
Comment (Sakshi-gupta-GL) [reply src/Pages/License/Generate/ProductDetails.jsx:104]: done

**Review by pinak-globallogic — APPROVED** [review]

**Review by arpitmanglik — APPROVED** [review]
